### PR TITLE
feat(authentication): resend verification email 

### PR DESCRIPTION
**Pull request by ChrisPdgn**
**What kind of change does this PR introduce?** 

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [x] No

**Other information:**
Resend verification email when creating a new user

### CODE REVIEW

File: modules/authentication/src/handlers/local.ts [hunk 1]
```diff
@@ -161,6 +161,19 @@ export class LocalHandlers implements IAuthenticationStrategy {
       this.verifyChangeEmail.bind(this),
     );
 
+    routingManager.route(
+      {
+        path: '/local/resend-verification',
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:166]: Make this `/local/new/send-verification` as it's not clear what sort of verification send this URI performs.
Update handler name to reflect this.

This should also only get registered whenever `this.sendVerificationEmail` is true.

File: modules/authentication/src/handlers/local.ts [hunk 2]
```diff
@@ -756,6 +773,45 @@ export class LocalHandlers implements IAuthenticationStrategy {
     return 'Email changed successfully';
   }
 
+  async resendVerificationEmail(
+    call: ParsedRouterRequest,
+  ): Promise<UnparsedRouterResponse> {
+    if (!this.sendVerificationEmail) return 'Configuration not set to send email';
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:779]: We shouldn't be registering this route in that case.

File: modules/authentication/src/handlers/local.ts [hunk 3]
```diff
@@ -756,6 +773,45 @@ export class LocalHandlers implements IAuthenticationStrategy {
...
+    let { email } = call.request.params;
+    if (AuthUtils.invalidEmailAddress(email)) {
+      throw new GrpcError(status.INVALID_ARGUMENT, 'Invalid email address provided');
+    }
+    email = email.toLowerCase();
+    const user: User | null = await User.getInstance().findOne({
+      email: email,
+    });
+    if (isNil(user)) throw new GrpcError(status.NOT_FOUND, 'User not found');
+    const verificationToken: Token | null = await Token.getInstance().findOne({
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:790]: We should probably recreate these instead, replacing any existing ones of that type for target user.
What if we let these expire down the line?
What if token creation failed for whatever reason, in which case we're stuck with a non-verifiable user.

File: modules/authentication/src/handlers/local.ts [hunk 4]
```diff
@@ -756,6 +773,45 @@ export class LocalHandlers implements IAuthenticationStrategy {
...
+    });
+    if (isNil(verificationToken)) {
+      throw new GrpcError(status.NOT_FOUND, 'Verification email token does not exist');
+    }
+    const serverConfig = await this.grpcSdk.config.get('router');
+    const url = serverConfig.hostUrl;
+    const result = { verificationToken, hostUrl: url };
+    const link = `${result.hostUrl}/hook/authentication/verify-email/${result.verificationToken.token}`;
+    await this.emailModule
+      .sendEmail('ChangeEmailVerification', {
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/local.ts:802]: you're using the change email verification instead of the email verification template

**Review by kkopanidis — CHANGES_REQUESTED** [review]

File: modules/authentication/src/handlers/local.ts [hunk 5]
```diff
@@ -756,6 +773,45 @@ export class LocalHandlers implements IAuthenticationStrategy {
...
+        email: email,
+        sender: 'no-reply',
+        variables: {
+          link,
+        },
+      })
+      .catch(e => {
+        ConduitGrpcSdk.Logger.error(e);
+      });
+    return 'Verification email resent';
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:812]: Make this `Verification code sent`.
We shouldn't treat this as a resend route/function.

**Review by kon14 — CHANGES_REQUESTED** [review]

File: modules/authentication/src/handlers/local.ts [hunk 6]
```diff
@@ -161,18 +161,20 @@ export class LocalHandlers implements IAuthenticationStrategy {
...
     );
 
-    routingManager.route(
-      {
-        path: '/local/resend-verification',
-        action: ConduitRouteActions.POST,
-        description: `Used to resend email verification after new user is created`,
-        bodyParams: {
-          email: ConduitString.Required,
+    if (this.sendVerificationEmail) {
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:164]: This should only register whenever `this.sendVerificationEmail` is true.
Move this into the relevant block above.

File: modules/authentication/src/handlers/local.ts [hunk 7]
```diff
@@ -756,6 +775,60 @@ export class LocalHandlers implements IAuthenticationStrategy {
...
+    let { email } = call.request.params;
+    if (AuthUtils.invalidEmailAddress(email)) {
+      throw new GrpcError(status.INVALID_ARGUMENT, 'Invalid email address provided');
+    }
+    email = email.toLowerCase();
+    const user: User | null = await User.getInstance().findOne({
+      email: email,
+    });
+    if (isNil(user)) throw new GrpcError(status.NOT_FOUND, 'User not found');
+    if (user.isVerified) return 'User is already verified';
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:790]: This should throw an error instead.

**Review by kon14 — CHANGES_REQUESTED** [review]

File: modules/authentication/src/handlers/local.ts [hunk 8]
```diff
@@ -787,21 +785,13 @@ export class LocalHandlers implements IAuthenticationStrategy {
       email: email,
     });
     if (isNil(user)) throw new GrpcError(status.NOT_FOUND, 'User not found');
-    if (user.isVerified) return 'User is already verified';
+    if (user.isVerified)
+      throw new GrpcError(status.FAILED_PRECONDITION, 'User already exists');
```
Comment (kon14) [inline modules/authentication/src/handlers/local.ts:789]: `User is already verified` was fine. I only requested that the return be changed to a throw.

**Review by kon14 — CHANGES_REQUESTED** [review]

**Review by kon14 — APPROVED** [review]

**Review by kkopanidis — APPROVED** [review]

**Pull request merged**